### PR TITLE
DO NOT MERGE - LAU -199 Doucments-delete end-point

### PR DESCRIPTION
**Pull request by RababRabab**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/LAU-284

### Change description ###

Addition of a new POST /documents/delete endpoint

**Does this PR introduce a breaking change?** (check one with "x")

[ ] Yes
[x] No

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/dm/service/AuditedStoredDocumentOperationsServiceTests.java [hunk 1]
```diff
@@ -187,5 +189,18 @@ public void testBulkUpdateDocumentWithError() {
...
+        storedDocuments.addAll(Stream.of(TestUtil.STORED_DOCUMENT).collect(Collectors.toList()));
+
+        CaseDocumentsDeletionResults caseDocumentsDeletionResults =
+            auditedStoredDocumentOperationsService.deleteCaseStoredDocuments(storedDocuments);
+
+        assertThat(caseDocumentsDeletionResults.getCaseDocumentsFound().equals(storedDocuments.size()));
+        assertThat(caseDocumentsDeletionResults.getMarkedForDeletion().equals(storedDocuments.size()));
+    }
+
 }
```
Comment (yogesh-hullatti) [inline src/test/java/uk/gov/hmcts/dm/service/AuditedStoredDocumentOperationsServiceTests.java:205]: Can we have one more Test where the CaseDocumentsFound don't match the MarkedForDeletion?

File: src/test/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteControllerTests.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        DeleteCaseDocumentsCommand deleteCaseDocumentsCommand = new DeleteCaseDocumentsCommand();
+
+        restActions
+            .withAuthorizedUser("userId")
+            .withAuthorizedService("ccd_data")
+            .post("/documents/delete", deleteCaseDocumentsCommand)
+            .andExpect(status().is4xxClientError());
+    }
+
+}
```
Comment (yogesh-hullatti) [inline src/test/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteControllerTests.java:68]: This seems to be Mocked Integration tests. We will need functional tests which will create required test data and then call different Delete scenarios. Something similar to what we have for other endpoints here - https://github.com/hmcts/document-management-store-app/tree/master/src/functionalTest/groovy/uk/gov/hmcts/dm/functional

File: src/main/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteController.java [hunk 3]
```diff
@@ -41,6 +59,55 @@
...
+                invalidRequestException.getMessage(),
+                invalidRequestException
+            );
+            return new ResponseEntity<>(null, BAD_REQUEST);
+        } catch (final Exception exception) {
+            log.error("deleteCaseDocuments API call failed due to error - {}",
+                exception.getMessage(),
+                exception
+            );
+            return new ResponseEntity<>(null, INTERNAL_SERVER_ERROR);
```
Comment (yogesh-hullatti) [inline src/main/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteController.java:109]: Can we please remove the code which returns the 500 Error Response. We had similar Response which after latest review and been removed. Please refer to this,
https://github.com/hmcts/document-management-store-app/blob/31c105cb3659bc9da21fc5dbb0ca898f5b1da6b4/src/main/java/uk/gov/hmcts/dm/controller/StoredDocumentController.java#L132
Returning 500 at times hides the actual cause of the error and results in Wrong Stats in the Dynatrace.

File: src/main/java/uk/gov/hmcts/dm/service/Constants.java [hunk 4]
```diff
@@ -4,4 +4,12 @@
 
     String IS_ADMIN = "isAdmin";
 
+    Boolean FALSE = false;
+
+    String CASE_REF_REGEX = "^\\d{16}$";
```
Comment (yogesh-hullatti) [inline src/main/java/uk/gov/hmcts/dm/service/Constants.java:9]: I am not sure if we have similar 16 digits validation for CaseRef across all applications? If, yes. Then we can leave this number to 16.

File: src/test/java/uk/gov/hmcts/dm/service/AuditedStoredDocumentOperationsServiceTests.java (hunk 1 above)
Comment (RababRabab) [reply src/test/java/uk/gov/hmcts/dm/service/AuditedStoredDocumentOperationsServiceTests.java:217]: Unlikely to happen in real life situation, but added a test for this scenario.

File: src/test/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteControllerTests.java (hunk 2 above)
Comment (RababRabab) [reply src/test/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteControllerTests.java:68]: Added 8 functional tests under DeleteDocumentsIT.groovy.

File: src/main/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteController.java (hunk 3 above)
Comment (RababRabab) [reply src/main/java/uk/gov/hmcts/dm/controller/StoredDocumentDeleteController.java]: Removed 500 error as requested.

File: src/main/java/uk/gov/hmcts/dm/service/Constants.java (hunk 4 above)
Comment (RababRabab) [reply src/main/java/uk/gov/hmcts/dm/service/Constants.java:9]: This is specified as part of the end-point design, case references are 16 digist long:
https://tools.hmcts.net/confluence/pages/viewpage.action?pageId=1545344961&__ncforminfo=SrXn1AZ7tEZMTraGYHoZrDgvdLA2l7NIkdJFSu8_8di0rnFhv5-CzAMBFusZ-bWRTKI7PtR_T4FGgtIxRx3hdjH5w188wYVr

**Review by matt-k1998 — APPROVED** [review]

**Pull request merged**